src/base/QXmppJingleData.cpp

```diff
@@ -3224,3 +3224,319 @@ std::optional<QXmppJingleMessageInitiationElement::Type> QXmppJingleMessageIniti
+///
+
+///
+/// \class QXmppCallInviteElement
+///
+/// \brief The QXmppCallInviteElement class represents a Call Invite
+/// element as specified by \xep{0482, Call Invites}.
+///
+/// \ingroup Stanzas
+///
+/// \since QXmpp 1.6
+///
+
+///
+/// \brief Constructs a Call Invite element.
+///
+QXmppCallInviteElement::QXmppCallInviteElement()
+    : d(new QXmppCallInviteElementPrivate())
+{
+}
+
+///
+/// Returns the Call Invite element type.
+///
+QXmppCallInviteElement::Type QXmppCallInviteElement::type() const
+{
+    return d->type;
+}
+
+///
+/// Sets the Call Invite element type.
+///
+void QXmppCallInviteElement::setType(Type type)
+{
+    d->type = type;
+}
+
+///
+/// Returns the Call Invite element id.
+///
+QString QXmppCallInviteElement::id() const
+{
+    return d->id;
+}
+
+///
+/// Sets the Call Invite element id.
+///
+void QXmppCallInviteElement::setId(const QString &id)
+{
+    d->id = id;
+}
+
+///
+/// Returns the Call Invite element audio flag.
+///
+bool QXmppCallInviteElement::audio() const
+{
+    return d->audio;
+}
+
+///
+/// Sets the Call Invite element audio flag.
+///
+void QXmppCallInviteElement::setAudio(bool audio)
+{
+    d->audio = audio;
+}
+
+///
+/// Returns the Call Invite element video flag.
+///
+bool QXmppCallInviteElement::video() const
+{
+    return d->video;
+}
+
+///
+/// Sets the Call Invite element video flag.
+///
+void QXmppCallInviteElement::setVideo(bool video)
+{
+    d->video = video;
+}
+
+///
+/// Returns a possible Call Invite element "jingle" sub element.
+///
+std::optional<QXmppCallInviteElement::Jingle> QXmppCallInviteElement::jingle() const
+{
+    return d->jingle;
+}
+
+///
+/// Sets a possible Call Invite "jingle" sub element.
+///
+void QXmppCallInviteElement::setJingle(std::optional<Jingle> jingle)
+{
+    d->jingle = jingle;
+}
+
+///
+/// Returns possible Call Invite "external" sub elements.
+///
+std::optional<QVector<QXmppCallInviteElement::External>> QXmppCallInviteElement::external() const
+{
+    return d->external;
+}
+
+///
+/// Sets possible Call Invite "external" sub elements.
+///
+void QXmppCallInviteElement::setExternal(std::optional<QVector<External>> external)
+{
+    d->external = external;
+}
+
+/// \cond
+void QXmppCallInviteElement::parse(const QDomElement &element)
+{
+    std::optional<Type> type { stringToCallInviteElementType(element.nodeName()) };
+
+    if (!type) {
+        return;
+    }
+
+    d->type = type.value();
+    d->id = element.attribute(QStringLiteral("id"));
+
+    switch (d->type) {
+    case Type::Invite:
+        d->audio = element.attribute(QStringLiteral("audio"), QStringLiteral("true")) == QStringLiteral("true") ? true : false;
+        d->video = element.attribute(QStringLiteral("video"), QStringLiteral("false")) == QStringLiteral("true") ? true : false;
+        // fall through
+    case Type::Accept: {
+        if (const auto &jingleElement = element.firstChildElement("jingle"); !jingleElement.isNull()) {
```

Comment:
No const ref to temporary here please

Reply:
It's not about thread-safety here.
* `firstChildElement()` returns by value (`QDomElement` not `const QDomElement &`).
* taking `const &` of the return value creates a reference to a temporary object (and I guess that's probably not what you wanted to do?)
* C++ says that the lifetimes of temporaries is extended in such cases though, so you won't get a dangling reference and the code works
* still i think that this is contra-intuitive and it looks like `firstChildElement()` would return a `const &`
* taking `const &` here is not faster, both variants generate the same binary
* there's nothing wrong with using `auto jingleElement = ...` (or am I overlooking something?)